src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+
+                        if (resourceExecutingContext.Result != null)
                         {
-                            Canceled = true,
-                            Result = _resourceExecutingContext.Result,
-                        };
+                            _resourceExecutedContext = new ResourceExecutedContext(resourceExecutingContext, _filters)
+                            {
+                                Canceled = true,
+                                Result = _resourceExecutingContext.Result,
+                            };
+
+                            goto case State.ResourceShortCircuit;
+                        }
+
+                        var task = InvokeNextResourceFilter();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ResourceSyncEnd;
+                            return task;
+                        }
+
+                        goto case State.ResourceSyncEnd;
                     }
-                    else
+
+                case State.ResourceSyncEnd:
+                    {
+                        var filter = (IResourceFilter)state;
+                        var resourceExecutedContext = _resourceExecutedContext;
+
+                        _diagnosticSource.BeforeOnResourceExecuted(resourceExecutedContext, filter);
+
+                        filter.OnResourceExecuted(resourceExecutedContext);
+
+                        _diagnosticSource.AfterOnResourceExecuted(resourceExecutedContext, filter);
+
+                        goto case State.ResourceEnd;
+                    }
+
+                case State.ResourceShortCircuit:
                     {
-                        await InvokeResourceFilterAsync();
                         Debug.Assert(_resourceExecutedContext != null);
 
-                        _diagnosticSource.BeforeOnResourceExecuted(_resourceExecutedContext, item.Filter);
+                        _logger.ResourceFilterShortCircuited((IFilterMetadata)state);
+
+                        var task = InvokeResultAsync(_resourceExecutingContext.Result);
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ResourceEnd;
+                            return task;
+                        }
+
+                        goto case State.ResourceEnd;
+                    }
+
+                case State.ResourceInside:
+                    {
+                        goto case State.ExceptionBegin;
+                    }
+
+                case State.ResourceEnd:
+                    {
+                        if (scope == Scope.Resource)
+                        {
+                            isCompleted = true;
+                            return TaskCache.CompletedTask;
+                        }
+
+                        Debug.Assert(scope == Scope.Invoker);
+                        Rethrow(_resourceExecutedContext);
```

Comment:
This makes it like it rethrows an exception unconditionally, it will be nice to be a bit more verbose and change the name to reflect that it will only throw if the previous scope ended in an unhandled exception. Something like RethrowIfUnhandledException